geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/GroupBy.scala

```diff
@@ -0,0 +1,113 @@
+
+case class GroupBy[T](attribute: Int, exampleStat: Stat)(implicit ct: ClassTag[T]) extends Stat {
+  override type S = GroupBy[T]
+
+  val groupedStats: mutable.HashMap[T, Stat] = mutable.HashMap[T, Stat]()
+
+  /**
+    * Compute statistics based upon the given simple feature.
+    * This method will be called for every SimpleFeature a query returns.
+    *
+    * @param sf feature to evaluate
+    */
+  override def observe(sf: SimpleFeature): Unit = {
+    val key = sf.getAttribute(attribute).asInstanceOf[T]
+    groupedStats.get(key) match {
+      case Some(groupedStat) => groupedStat.observe(sf)
+      case None              => val newStat = exampleStat.newcopy
+        newStat.observe(sf)
+        groupedStats.update(key, newStat)
+    }
+  }
+
+  /**
+    * Tries to remove the given simple feature from the compiled statistics.
+    * Note: may not be possible to un-observe a feature, in which case this method will
+    * have no effect.
+    *
+    * @param sf feature to un-evaluate
+    */
+  override def unobserve(sf: SimpleFeature): Unit = {
+    val key = sf.getAttribute(attribute).asInstanceOf[T]
+    groupedStats.get(key) match { case Some(groupedStat) => groupedStat.unobserve(sf) }
+  }
+
+  /**
+    * Add another stat to this stat. Avoids allocating another object.
+    *
+    * @param other the other stat to add
+    */
+  override def +=(other: GroupBy[T]): Unit = {
+    other.groupedStats.map { case (key, stat) =>
+      groupedStats.get(key) match {
+        case Some(groupedStat) => groupedStat += stat
+        case None              => groupedStats.put(key, stat)
+      }
+    }
+  }
+
+  /**
+    * Combine two stats into a new stat
+    *
+    * @param other the other stat to add
+    */
+  override def +(other: GroupBy[T]): GroupBy[T] = {
+    val newGB = new GroupBy[T](attribute, exampleStat.newcopy)
+    newGB += this
+    newGB += other
+    newGB
+  }
+
+  /**
+    * Returns a json representation of the stat
+    *
+    * @return stat as a json string
+    */
+  override def toJson: String = {
+    // TODO: Provide summary of all items in all groups together. e.g. GroupBy("cat",Count()) should have a total count as well as a per category count.
+    groupedStats.map{ case (key, stat) => "{ \"" + key + "\" : " + stat.toJson + "}" }.mkString("[",",","]")
+  }
+
+  /**
+    * Necessary method used by the StatIterator. Indicates if the stat has any values or not
+    *
+    * @return true if stat contains values
+    */
+  override def isEmpty: Boolean = groupedStats.isEmpty
+
+  /**
+    * Compares the two stats for equivalence. We don't use standard 'equals' as it gets messy with
+    * mutable state and hash codes
+    *
+    * @param other other stat to compare
+    * @return true if equals
+    */
+  override def isEquivalent(other: Stat): Boolean = {
+    other match {
+      case other: GroupBy[T] => !groupedStats.map{ case (key, stat) => other.groupedStats.get(key) == stat }.exists(p => p == false)
```

Comment:
is `==` appropriate here?  I would think a delegated per group `isEquivalent` would be appropriate